<a href="https://colab.research.google.com/github/Kevin-AngelQD/padp/blob/main/Actividad_Calificada_07_Datos_Temporales_271758.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | Kevin Angel Quiñonez Dominguez |
| **Matrícula** |271758|

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [2]:
# Escribe tu código aquí

print(taxis.pickup.dtype)
print(taxis.dropoff.dtype)#Ambas son object antes de su transformacion

object
object


In [3]:
print('Faltantes de pickup antes de convertir:', taxis.pickup.isna().sum())
print('Faltantes de dropoff antes de convertir:',taxis.dropoff.isna().sum())#

taxis.pickup = pd.to_datetime(taxis.pickup, format='%Y-%m-%d %H:%M:%S')
taxis.dropoff = pd.to_datetime(taxis.dropoff, format='%Y-%m-%d %H:%M:%S')

print('NaT de pickup después de convertir    :', taxis.pickup.isna().sum())
print('NaT de dropoff después de convertir    :',taxis.dropoff.isna().sum()) #No genero NaT

Faltantes de pickup antes de convertir: 0
Faltantes de dropoff antes de convertir: 0
NaT de pickup después de convertir    : 0
NaT de dropoff después de convertir    : 0


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [4]:
# Escribe tu código aquí

print('Fecha mínima de pickup:', taxis.pickup.min())
print('Fecha máxima de pickup:', taxis.pickup.max())
print('Fecha mínima de dropoff:', taxis.dropoff.min())
print('Fecha máxima de dropoff:', taxis.dropoff.max())


taxis_sin_marzo = taxis[(taxis['pickup'] < '2019-03-01 00:00:00') | (taxis['pickup'] > '2019-03-31 23:59:59')] #Se uso hasta el ultimo segundo en la comparativa para asegurar
print('Viajes empezados fuera de marzo: \n',taxis_sin_marzo)
print('Total de viajes empezados fuera de marzo: ',taxis_sin_marzo.shape[0])#El resultado confirma la fecha minima de pickup

taxis_dropoff_abril=taxis[(taxis['pickup'] >= '2019-03-01 00:00:00') & (taxis['dropoff'] > '2019-03-31 23:59:59')]
#print('Total de viajes realizados fuera de marzo 2019: ', taxis[~((taxis['pickup'].dt.month == 3) & (taxis['pickup'].dt.year == 2019))]) #codigo alternativo
print('\nViajes empezados en marzo que terminaron en abril:\n ',taxis_dropoff_abril)
print('Total de viajes terminados fuera de marzo: ',taxis_dropoff_abril.shape[0])#El resultado confirma la fecha maxima de dropoff


Fecha mínima de pickup: 2019-02-28 23:29:03
Fecha máxima de pickup: 2019-03-31 23:43:45
Fecha mínima de dropoff: 2019-02-28 23:32:35
Fecha máxima de dropoff: 2019-04-01 00:13:58
Viajes empezados fuera de marzo: 
                   pickup             dropoff  passengers  distance  fare  tip  \
6203 2019-02-28 23:29:03 2019-02-28 23:32:35           1       0.9   5.0  0.0   

      tolls  total  color payment  pickup_zone                   dropoff_zone  \
6203    0.0    6.3  green    cash  Old Astoria  Long Island City/Queens Plaza   

     pickup_borough dropoff_borough  
6203         Queens          Queens  
Total de viajes empezados fuera de marzo:  1

Viajes empezados en marzo que terminaron en abril:
                   pickup             dropoff  passengers  distance  fare  tip  \
591 2019-03-31 23:43:45 2019-04-01 00:13:58           5     12.25  37.0  0.0   

     tolls  total   color payment        pickup_zone  \
591    0.0   40.8  yellow    cash  LaGuardia Airport   

            

### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [5]:
# Escribe tu código aquí

import datetime as dt
viajes_10_marzo = taxis[(taxis['pickup'] >= '2019-03-10 00:00:00') & (taxis['pickup'] < '2019-03-11 00:00:00')]
# Agrupamos por la hora de esa misma variable y contamos con .size()
print('Viajes por hora el 10 de marzo:\n', viajes_10_marzo.groupby(viajes_10_marzo['pickup'].dt.hour).size())

print(taxis['pickup'].dt.tz_localize('America/New_York', nonexistent='NaT').isna().sum())

Viajes por hora el 10 de marzo:
 pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
dtype: int64
0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**
De manera momentanea tomando en cuenta los resultados, ninguno. A lo mucho que habia pocos datos fuera de marzo pero talvez es el hecho de que el dataset tiene enfoque en Marzo. No hubo NaTs antes ni despues de formatear las fechas

El resultado de 1.3 indica que no hubo conflictos de horario y por lo tanto los datos estan limpios en ese aspecto.

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [6]:
# Escribe tu código aquí
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60
taxis['duracion_min'].describe()

,duracion_min
count,6433.000000
mean,14.349617
std,11.643892
min,0.000000
25%,6.500000
50%,10.900000
75%,18.516667
max,107.666667


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [7]:
# Escribe tu código aquí
print('Viajes con duracion negativa: ', taxis[taxis['duracion'] < pd.Timedelta(0)].shape[0])
print('Viajes con duracion cero: ', taxis[taxis['duracion'] == pd.Timedelta(0)].shape[0])
print('Viajes con duracion menor a 1 minuto: ', taxis[taxis['duracion_min'] < 1].shape[0])
print('Viajes con duracion mayor a 3 horas: ', taxis[taxis['duracion_min'] > 180].shape[0])
taxis[taxis['duracion_min'] == 0].loc[:,['pickup','dropoff','distance','fare','total']]

#Se necesita contexto al respecto de esos viajes con cobros y duracion de cero segundos

Viajes con duracion negativa:  0
Viajes con duracion cero:  6
Viajes con duracion menor a 1 minuto:  72
Viajes con duracion mayor a 3 horas:  0


,pickup,dropoff,distance,fare,total
1690,2019-03-22 06:24:14,2019-03-22 06:24:14,0.0,72.0,72.0
5493,2019-03-06 12:14:00,2019-03-06 12:14:00,0.0,10.0,10.0
5624,2019-03-11 14:04:50,2019-03-11 14:04:50,0.0,2.5,3.3
5638,2019-03-13 12:22:00,2019-03-13 12:22:00,0.0,10.0,10.0
6083,2019-03-19 15:34:00,2019-03-19 15:34:00,0.0,5.0,5.0
6356,2019-03-01 16:58:23,2019-03-01 16:58:23,0.0,3.0,4.8


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [8]:
# Escribe tu código aquí

import numpy as np

#Calculamos la velocidad base para todos los viajes
velocidad_base = taxis["distance"] / (taxis["duracion_min"] / 60)

#Aplicamos .where(): Conserva el valor SI la duración es > 0, de lo contrario pone NaN
# Nota: Si la distancia es 0 y la duración > 0, dará 0.0 automáticamente.
taxis["velocidad_mph"] = velocidad_base.where(taxis["duracion_min"] > 0, np.nan)

#Contar viajes con velocidad > 60 mph (Pandas ignora los NaN automáticamente aquí)
print("Viajes con velocidad mayor a 60 mph: ", taxis[taxis["velocidad_mph"] > 60].shape[0])

#Mostrar los 5 viajes con mayor velocidad usando las columnas solicitadas
columnas_reporte = ["pickup", "dropoff", "duracion_min", "distance", "velocidad_mph", "fare"]

print("\nLos 5 viajes con mayor velocidad:")
print(taxis.sort_values(by="velocidad_mph", ascending=False)[columnas_reporte].head(5))

Viajes con velocidad mayor a 60 mph:  11

Los 5 viajes con mayor velocidad:
                  pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

a) Desconozco al respecto de este medio, pero puede ser que sea la continuacion de otro viaje, es decir, lo que sucede que x viaje normal sucede y acto seguido un viaje corto al verdadero destino, entre otras cosas puede ser condiciones climatologicas que afectan caminar, personas de la tercera edad, viajes por seguridad en la noche o incluso transporte de equipaje o muebles lo que hace viable un taxi por menos de un minuto. En cuanto a la velocidad, lo viable seria informarse al respecto si puede ser un error de captura de la velocidad.

b)Mi umbral de decision seria:

- La duracion del viaje mientras sea superior a un minuto,  ya que puede ser un error de captura o viajes de longitud insignificante u errores. Tampoco viajes de 3 horas o mas.
- Velocidad inferior a 100 mph, cualquier fila con velocidad superior a eso podria ser un error de captura ya que esa velocidad podria poner en riesgo a pasajeros e infrigir limites de velocidad.
- En periodos no habria limites, ya que todas las fechas y horas son viables para viajes en taxi, a menos que por algun motivo se analice en ciertos horarios y/o dias entonces se tendrian que excluir datos.

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [9]:
# Escribe tu código aquí

taxis['hora'] = taxis['pickup'].dt.hour
taxis['dia_num'] = taxis['pickup'].dt.dayofweek #No se usa iso para asegurar que el lunes sea 0
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()
taxis['fin_semana'] = taxis['dia_num'].isin([5,6]) #Marca los sabados y domingos bajo una condicional
taxis['fecha'] = taxis['pickup'].dt.date
taxis['cruza_medianoche'] = taxis['pickup'].dt.date != taxis['dropoff'].dt.date

print('Viajes que cruzan la medianoche: \n', taxis[taxis['cruza_medianoche'] == True].shape[0])

Viajes que cruzan la medianoche: 
 50


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [10]:
# Escribe tu código aquí

print('Viajes por hora y mediana', taxis.groupby('hora').agg({'duracion_min': ['count', 'median']}))
#Se pueden confirmar los datos checando la lista de viaje por horas
print('Hora con mas viajes: \n', taxis.groupby('hora').size().idxmax())
print('Hora con menos viajes: \n', taxis.groupby('hora').size().idxmin())

Viajes por hora y mediana      duracion_min           
            count     median
hora                        
0             205  10.733333
1             110   9.633333
2             101   9.183333
3              67   9.583333
4              57  10.500000
5              51   8.700000
6             142   7.833333
7             221   9.583333
8             313  11.316667
9             319  11.616667
10            325  11.883333
11            295  12.033333
12            334  12.775000
13            316  10.558333
14            360  11.841667
15            330  12.058333
16            336  11.583333
17            388  11.975000
18            417  11.100000
19            406  10.650000
20            368   9.258333
21            355  11.083333
22            321   9.816667
23            296  10.575000
Hora con mas viajes: 
 18
Hora con menos viajes: 
 5


### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [11]:
# Escribe tu código aquí

print('Viajes por dia y mediana', taxis.groupby(['dia_num', 'dia_nombre']).agg({'duracion_min': ['count', 'median'], 'fare': ['count', 'median']}))
print('Promedio de viajes por fecha', taxis.groupby('dia_num').size() / taxis['fecha'].nunique())


Viajes por dia y mediana                    duracion_min             fare       
                          count     median count median
dia_num dia_nombre                                     
0       Monday              708  10.541667   708    9.5
1       Tuesday             825  11.216667   825    9.5
2       Wednesday           966  11.291667   966    9.5
3       Thursday            905  11.766667   905    9.5
4       Friday             1115  11.150000  1115    9.5
5       Saturday           1046  10.466667  1046    9.0
6       Sunday              868   9.983333   868    9.0
Promedio de viajes por fecha dia_num
0    22.12500
1    25.78125
2    30.18750
3    28.28125
4    34.84375
5    32.68750
6    27.12500
dtype: float64


### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

a) No coinciden, que tenga mas viajes no significa que los viajes sean mas largos como tal, puede ser por el hecho de que hay mas viajes que son mas cortos lo que mueve la mediana hacia abajo.

b) Esto se debe a que puede haber mas dias en particular que otros (cinco lunes, cuatro martes, etc) dividirlo por las fechas arregla este desbalance dando los valores tipicos de dichas fechas. Tambien arregla desbalances por dias festivos y/o fallas en el sistema.

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [12]:
# Escribe tu código aquí

#Creacion del dataset limpio
filtro_duracion = (taxis["duracion_min"] < 1) | (taxis["duracion_min"] > 180)
filtro_velocidad = (taxis["velocidad_mph"] > 100)
taxis_limpio = taxis[~(filtro_duracion | filtro_velocidad)]



print('Viajes excluidos por duracion menor a un minuto o superior a 3 horas: ', taxis[filtro_duracion].shape[0])

print('Viajes excluidos por velocidad superior a 100 MPH: ', taxis[filtro_velocidad].shape[0])
#Fun fact, sume esta linea con la superior dio 83 lineas crei que estaba haciendo algo mal en la eliminacion
#hasta que me di cuenta de que 73 de esos viajes podrian cumplir lo de las 11 lineas.

print('Viajes excluidos en total: ', taxis.shape[0] - taxis_limpio.shape[0])

#En estas no se uso el filtro debido a la diferencia de tamaño, lo cual se puede arreglar con loc pero decidi usar el codigo entero
print('Viajes de duracion menor a un minuto o superior a 3 horas en taxis_limpio: ',
      taxis_limpio[(taxis_limpio['duracion_min'] < 1) | (taxis_limpio['duracion_min'] > 180)].shape[0])

print('Viajes de velocidad superior a 100 MPH en taxis_limpio: ',taxis_limpio[taxis_limpio["velocidad_mph"] > 100].shape[0])

print("Total de viajes originales: ",taxis.shape[0])
print("Total de viajes despues de filtrado: ",taxis_limpio.shape[0])
print("Total de viajes excluidos: ",len(taxis) - len(taxis_limpio))



Viajes excluidos por duracion menor a un minuto o superior a 3 horas:  72
Viajes excluidos por velocidad superior a 100 MPH:  11
Viajes excluidos en total:  73
Viajes de duracion menor a un minuto o superior a 3 horas en taxis_limpio:  0
Viajes de velocidad superior a 100 MPH en taxis_limpio:  0
Total de viajes originales:  6433
Total de viajes despues de filtrado:  6360
Total de viajes excluidos:  73


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [13]:
# Escribe tu código aquí
#print("Viajes en dia habil: \n",taxis_limpio[(taxis_limpio['hora'] >= 7) & (taxis_limpio['hora'] <= 9) & (taxis_limpio['fin_semana'] == False)].describe())

#print("Viajes en dia inhabil: \n",taxis_limpio[(taxis_limpio['hora'] >= 7) & (taxis_limpio['hora'] <= 9) & (taxis_limpio['fin_semana'] == True)].describe())


#Definimos la condición de día hábil e inhabil
filtro_dia_habil = (
    (taxis_limpio["hora"] >= 7)
    & (taxis_limpio["hora"] <= 9)
    & (taxis_limpio["fin_semana"] == False)
)

filtro_dia_inhabil = (
    (taxis_limpio["hora"] >= 7)
    & (taxis_limpio["hora"] <= 9)
    & (taxis_limpio["fin_semana"] == True)
)

#Filtramos el DataFrame
viajes_habiles = taxis_limpio[filtro_dia_habil]
viajes_inhabiles = taxis_limpio[filtro_dia_inhabil]

#Calculamos el total de filas y las medianas específicas
medianas_habiles = viajes_habiles[["duracion_min", "velocidad_mph"]].agg("median")
medianas_inhabiles = viajes_inhabiles[["duracion_min", "velocidad_mph"]].agg("median")

print("=== Métricas de Viajes en Día Hábil ===")
print(f"Total de viajes: " ,viajes_habiles.shape[0])
print(f"Mediana de duración (min): {medianas_habiles['duracion_min']}")
print(f"Mediana de velocidad (MPH): {medianas_habiles['velocidad_mph']}\n")

print("=== Métricas de Viajes en Día Inhábil ===")
print(f"Total de viajes: ", viajes_inhabiles.shape[0])
print(f"Mediana de duración (min): {medianas_inhabiles['duracion_min']}")
print(f"Mediana de velocidad (MPH): {medianas_inhabiles['velocidad_mph']}")

=== Métricas de Viajes en Día Hábil ===
Total de viajes:  683
Mediana de duración (min): 12.033333333333333
Mediana de velocidad (MPH): 9.097978227060652

=== Métricas de Viajes en Día Inhábil ===
Total de viajes:  161
Mediana de duración (min): 7.766666666666667
Mediana de velocidad (MPH): 13.47305389221557


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

- **Conversión:** Se convirtieron las fechas de pickup y dropoff en el formato de año, mes y dia. No hubo errores ni datos faltantes en dichas columnas.
- **Zona horaria:** Estaban en horario de New York. Esto debido a que los horarios no chocaron con las horas al darles dicho formato, no se generaron inconsistencias.
- **Criterio de exclusión:** Se aplicaron las condiciones siguientes para descartar las filas. Esto debido a que cualquier viaje de muy pocos segundos puede ser un error de captura. Cualquier viaje superior a 3 horas puede considerarse un error tambien, puede causar sesgos para el viaje promedio (y posiblemente el limite de tiempo para un viaje):

    1.   Duracion de menos de un minuto o superior a 3 horas
    2.   Cualquier viaje que supere 100 millas por hora

- **Resultado:** La cantidad de viajes originales era: 6433, se redujeron 73 viajes, quedaron 6360


**Hallazgo de 4.2 y limitaciones del análisis:** Con los datos de 4.2 podemos determinar que el trafico es reducido los fines de semana y la velocidad aumenta. Esto nos dice que las calles estan mas despejadas los fines de semana lo que permite viajes cortos y una mayor velocidad por viaje, el trafico excesivo entre semana reduce estos valores. En cuanto a limitaciones, una podria ser la velocidad real, la formula que realizamos es un estimado pero puede ser diferente en realidad, si un taxi recorre 3 millas en 20 minutos, la fórmula dirá que iba a 9 MPH. Sin embargo, la realidad de ese viaje pudo haber sido: el taxi avanzó a 30 MPH durante 5 minutos, pero estuvo totalmente detenido (0 MPH) durante 15 minutos en un semáforo o embotellamiento. La fórmula diluye por completo las aceleraciones y las paradas. Tomando en cuenta mi criterio el análisis queda estrictamente limitado al "flujo de tráfico urbano normalizado". No sirve para estudiar anomalías del servicio, emergencias extremas o incluso viajes largos.

---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

1. Indicar el formato deja fuera cualquier error posible de interpretacion. Si el mes fuera primero o el dia, en el que el primer espacio se podria ser incierto entre mes o dia podria haber tomado el formato americano (mes, dia, año) o el que se usa en Mexico (dia, mes, año) dando una fecha equivocada.

2. Un caos, tomando en cuenta mi criterio de viajes inferiores a un minuto, los hubiera eliminado dado a la diferencia de tiene UTC y New York. Suponiendo que son las 12PM en horario de New York en un pickup y se adapto con UTC, daria 16:00 en realidad y el dropoff a 12:30, dando una diferencia de -tres horas y media, elementos asi se eliminarian, e incluso si se inviertieran, generarian un sesgo destruyendo la verdadera precision. Una forma de detectar si hay un problema en el horario, en este caso seria la duracion cruda, es decir dropoff-pickup, comparar si es menor a cero y sumar todos las filas que cumplan esa caracteristica.

3. Seria inutilizable por que tendria corrupcion por los valores atipicos. Tomando en cuenta los viajes de cero segundos dividir por cero puede causar errores, aunque en este caso algoritmos de regresion lineal colapsarian al calcular algo como la perdida provocando errores del tipo NaN o OverflowError durante el entrenamiento. Los viajes cortos menores a un minuto podrian destrozar la varianza ya que creeran que es completamente normal un viaje de 10 segundos con 4000 millas por hora.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.